romancal/outlier_detection/utils.py

```diff
@@ -0,0 +1,314 @@
+        set to False.
+    """
+    in_memory = not input_models._on_disk
+    ngroups = len(input_models)
+    example_model = None
+
+    with input_models:
+        for i in range(len(input_models)):
+
+            model = input_models.borrow(i)
+            wht = build_driz_weight(
+                model,
+                # FIXME this was hard-coded to "ivm"
+                weight_type="ivm",
+                good_bits=good_bits,
+            )
+
+            if save_intermediate_results:
+                # write the model to file
+                _fileio.save_drizzled(model, make_output_path)
+
+            if i == 0:
+                input_shape = (ngroups,) + model.data.shape
+                dtype = model.data.dtype
+                computer = MedianComputer(input_shape, in_memory, buffer_size, dtype)
+                example_model = model
+                median_wcs = copy.deepcopy(model.meta.wcs)
+
+            weight_threshold = compute_weight_threshold(wht, maskpt)
+
+            data_copy = model.data.copy()
+            data_copy[wht < weight_threshold] = np.nan
+            computer.append(data_copy, i)
+
+            input_models.shelve(model, i, modify=True)
+            del data_copy, model
+
+    # Perform median combination on set of drizzled mosaics
+    median_data = computer.evaluate()
+
+    if save_intermediate_results:
+        _fileio.save_median(example_model, median_data, median_wcs, make_output_path)
+
+    return median_data, median_wcs
+
+
+def _flag_resampled_model_crs(
+    image,
+    median_data,
+    median_wcs,
+    snr1,
+    snr2,
+    scale1,
+    scale2,
+    backg,
+    save_intermediate_results,
+    make_output_path,
+):
+    blot = gwcs_blot(median_data, median_wcs, image.data.shape, image.meta.wcs, 1.0)
+
+    # Get background level of science data if it has not been subtracted, so it
+    # can be added into the level of the blotted data, which has been
+    # background-subtracted
+    if (
+        hasattr(image.meta, "background")
+        and image.meta.background.subtracted is False
+        and image.meta.background.level is not None
+    ):
+        backg = image.meta.background.level.value
+        log.debug(
+            f"Adding background level {image.meta.background.level} to blotted image"
+        )
+
+    cr_mask = flag_resampled_crs(
+        image.data.value, image.err.value, blot, snr1, snr2, scale1, scale2, backg
+    )
+
+    # update the dq flags in-place
+    image.dq |= cr_mask * np.uint32(pixel.DO_NOT_USE | pixel.OUTLIER)
+    log.info(f"{np.count_nonzero(cr_mask)} pixels marked as outliers")
+
+
+def _flag_model_crs(image, median_data, snr):
+    cr_mask = flag_crs(image.data.value, image.err.value, median_data, snr)
+
+    # Update dq array in-place
+    image.dq |= cr_mask * np.uint32(pixel.DO_NOT_USE | pixel.OUTLIER)
+
+    log.info(f"{np.count_nonzero(cr_mask)} pixels marked as outliers")
+
+
+def detect_outliers(
```

Comment:
probably a nitpick, but is the `detect_outliers` function really a utility that belongs in `utils`?

Reply:
I struggled to come up with a way to organize/name this so suggestions are welcome. Putting it in "imaging" (like in jwst) doesn't make sense here (as there's no non-imaging modes). Putting it in another "outlier_detection" submodule is overly redundant ("romancal.outlier_detection.outlier_detection.detect_outliers"). So I gave up and went with the standard "utils".

Reply:
It's not that many lines.  Is there a reason why it needs to be its own function, given that there is only one outlier detection mode for romancal?  That is, can you just embed the code into `OutlierDetectionStep`?

Reply:
I considered that. I believe the convention for romancal (as described by @mairanteodoro ) is to have the steps handle the stpipe-specific stuff and s much of the "meat" of the processing occur in a separate function/class/utility. If it's ok with you let's see what he thinks.

Reply:
sounds good it's definitely nbd either way as far as I'm concerned